homeassistant/components/konnected.py

```diff
@@ -0,0 +1,304 @@
+        vol.Exclusive(CONF_ZONE, 's_pin'): vol.Any(*ZONE_TO_PIN),
+        vol.Required(CONF_TYPE): DEVICE_CLASSES_SCHEMA,
+        vol.Optional(CONF_NAME): config_validation.string,
+    }), config_validation.has_at_least_one_key(CONF_PIN, CONF_ZONE)
+)
+
+_SWITCH_SCHEMA = vol.All(
+    vol.Schema({
+        vol.Exclusive(CONF_PIN, 'a_pin'): vol.Any(*PIN_TO_ZONE),
+        vol.Exclusive(CONF_ZONE, 'a_pin'): vol.Any(*ZONE_TO_PIN),
+        vol.Optional(CONF_NAME): config_validation.string,
+        vol.Optional('activation', default='high'):
+            vol.All(vol.Lower, vol.Any('high', 'low'))
+    }), config_validation.has_at_least_one_key(CONF_PIN, CONF_ZONE)
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema({
+            vol.Required('auth_token'): config_validation.string,
+            vol.Required(CONF_DEVICES): [{
+                vol.Required(CONF_ID, default=''): config_validation.string,
+                vol.Optional(CONF_SENSORS): [_SENSOR_SCHEMA],
+                vol.Optional(CONF_SWITCHES): [_SWITCH_SCHEMA],
+            }],
+        }),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+DEPENDENCIES = ['http', 'discovery']
+
+ENDPOINT_ROOT = '/api/konnected'
+UPDATE_ENDPOINT = (
+    ENDPOINT_ROOT +
+    r'/device/{device_id:[a-zA-Z0-9]+}/{pin_num:[0-9]}/{state:[01]}')
+
+
+async def async_setup(hass, config):
+    """Set up the Konnected platform."""
+    cfg = config.get(DOMAIN)
+    if cfg is None:
+        cfg = {}
+
+    auth_token = cfg.get('auth_token')
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {'auth_token': auth_token}
+
+    async def async_device_discovered(service, info):
+        """Call when a Konnected device has been discovered."""
+        _LOGGER.debug("Discovered a new Konnected device: %s", info)
+        host = info.get(CONF_HOST)
+        port = info.get(CONF_PORT)
+
+        device = KonnectedDevice(hass, host, port, cfg)
+        device.setup()
+
+    discovery.async_listen(
+        hass,
+        SERVICE_KONNECTED,
+        async_device_discovered)
+
+    hass.http.register_view(KonnectedView(auth_token))
+
+    return True
+
+
+class KonnectedDevice(object):
+    """A representation of a single Konnected device."""
+
+    def __init__(self, hass, host, port, config):
+        """Initialize the Konnected device."""
+        self.hass = hass
+        self.host = host
+        self.port = port
+        self.user_config = config
+
+        import konnected
+        self.client = konnected.Client(host, str(port))
+        self.status = self.client.get_status()
+        _LOGGER.info('Initialized Konnected device %s', self.device_id)
+
+    def setup(self):
+        """Set up a newly discovered Konnected device."""
+        user_config = self.config()
+        if user_config:
+            _LOGGER.debug('Configuring Konnected device %s', self.device_id)
+            self.save_data()
+            self.hass.async_add_job(self.sync_device)
```

Comment:
Now this is potentially an async call from a sync context. I know it's only called from inside the event loop right now, but every method that can be directly called from an async context must be prefixed `async_` and be either a coroutine or `@callback`. See https://developers.home-assistant.io/docs/en/asyncio_categorizing_functions.html

Reply:
This means that the `setup` function and `sync_device` should be annotated as async as well, right? I've categorized them both as callbacks because they don't have to wait for any result. Please let me know if I'm thinking about this wrong.